hop/robust_publisher.py

```diff
@@ -0,0 +1,640 @@
+                          overhead. Note that this size can become large is a lengthy network
+                          disruption prev ents messages from being sent; enough disk spacec should
+                          be available to cover this possibility for the expected message rate and
+                          duration of disruptions which may need to be handled.
+            poll_wait: The time the publisher should spend checking for receipt of each message
+                       directly after sending it. Tuning this parameter controls a tradeoff between
+                       low latency discovery of successful message delivery and throughput. If the
+                       time between sending messages is large compared to the latency for a message
+                       to be sent and for a confirmation of receipt to return, it is useful to
+                       increase this value so that the publisher will wait to discover that each
+                       message has been sent (in the success case) instead of sleeping and waiting
+                       for another message to send. If this value is 'too low' (much smaller than
+                       both the time for a mesage to be sent and acknowledged and the time for the
+                       next message to be ready for sending), the publisher will waste CPU time
+                       entering and exiting the internal function used to receive event
+                       notifications. If this value is too large (larger than or similar in size to
+                       the time between messages needing to be sent) throughput will be lost as time
+                       will be spent waiting to see if the previous message has been acknowledged
+                       which could be better spent getting the next message sent out. When in doubt,
+                       it is probably best to err on the side of choosing a small value.
+            kwargs: Any additional arguments to be passed to :meth:`hop.io.open <hop.io.open>`.
+
+        Raises:
+            OSError: If a journal file exists but cannot be read.
+            Runtime Error: If the contents of the journal file are corrupted.
+        """
+        super(RobustPublisher, self).__init__()
+
+        self._poll_wait = poll_wait
+
+        # read any messages left over from a previous run
+        self._journal = PublicationJournal(journal_path)
+        if self._journal.has_messages_to_send():
+            logger.info(f"Journal has {len(self._journal.messages_to_send)} "
+                        "left-over messages to send")
+
+        self._lock = threading.Lock()
+        self._cond = threading.Condition(self._lock)
+
+        dummy = io.Stream(auth=auth)
+        self._stream = dummy.open(url, "w", error_callback=PublicationJournal.error_callback,
+                                  **kwargs)
+
+        self._running = True
```

Comment:
I looked at this again, and renamed the variable `_should_stop`, since that is what it is really used to indicate, and is more along the lines of how I usually write this pattern.